File: proposed-projects/round-trip-experiment/reverse/20260909T185402Z_14873f98/20260910T102153Z_a4bbfe64/ported_implementation/src/gbnf.ts
import { Graph } from './grammarGraph/graph';
import { ParseState } from './grammarGraph/parseState';
import type { UnresolvedRule, ValidInput } from './grammarGraph/types';
import { buildRuleStack } from './grammarParser/buildRuleStack';
import { RulesBuilder } from './rulesBuilder';
import { GrammarParseError } from './utils/errors/grammarParseError';

export const GBNF = (input: string, initialString: ValidInput = ''): ParseState => {
  const grammar = typeof input === 'string' ? input : String(input);
  const { rules, symbolIds } = new RulesBuilder(grammar);
  if (rules.length === 0) {
    throw new GrammarParseError(grammar, 0, 'No rules were found');
  }
  if (!symbolIds.has('root')) {
    const available = JSON.stringify([...symbolIds.keys()]);
    throw new GrammarParseError(
      grammar,
      0,
      `Grammar does not contain a root symbol. Available symbols are: ${available}`,
    );
  }
  const rootId = symbolIds.get('root');

  const stackedRules: (UnresolvedRule[][] | undefined)[] = rules.map((rule) =>
    rule !== undefined ? buildRuleStack(rule) : undefined,
  );
  const graph = new Graph(grammar, stackedRules, rootId);
  return new ParseState(graph, graph.add(initialString));
};

File: packages/gbnf-experiment/data/20260909T032340Z_f01d5496/ported_implementation/gbnf/gbnf.py
import json

from .grammar_graph.graph import Graph
from .grammar_graph.parse_state import ParseState
from .grammar_graph.types import ValidInput
from .grammar_parser.build_rule_stack import build_rule_stack
from .rules_builder.rules_builder import RulesBuilder
from .utils.errors.grammar_parse_error import GrammarParseError


def GBNF(input_: str, initial_string: ValidInput = "") -> ParseState:
    grammar = input_ if isinstance(input_, str) else str(input_)
    builder = RulesBuilder(grammar)
    rules, symbol_ids = builder.rules, builder.symbol_ids
    if len(rules) == 0:
        raise GrammarParseError(grammar, 0, "No rules were found")
    if not symbol_ids.has("root"):
        raise GrammarParseError(
            grammar,
            0,
            "Grammar does not contain a root symbol. Available symbols are: "
            f"{json.dumps(list(symbol_ids.keys()))}",
        )
    root_id = symbol_ids.get("root")

    stacked_rules = [build_rule_stack(rule) for rule in rules]
    graph = Graph(grammar, stacked_rules, root_id)
    return ParseState(graph, graph.add(initial_string))
